samples/matter/common/src/watchdog/watchdog.cpp

```diff
@@ -0,0 +1,186 @@
+
+	/* Stop the Global Watchdog before further operations */
+	bool startAutomatically = false;
+	if (sEnabled) {
+		if (!Disable()) {
+			return false;
+		}
+		/* Watchdog was enabled so after doing all opreations for adding a new source start it automatically */
+		startAutomatically = true;
+
+		/* If the watchdog was enabled previously, restore all disabled sources */
+		sys_snode_t *node;
+		SYS_SLIST_FOR_EACH_NODE (&sWatchdogList, node) {
+			/* Restore all other Watchdog sources from the list */
+			if (node != &source) {
+				static_cast<WatchdogSource *>(node)->Install();
+			}
+		}
+	}
+
+	/* Try to install a new source */
+	if (source.Install()) {
+		/* Register timer in the list */
+		sys_slist_append(&sWatchdogList, &source);
+
+		/* Timer for feeding should be initialized only if feeding callback has been provided */
+		if (source.mCallback) {
+			k_timer_init(&source.mTimer, &WatchdogSource::TimerTimeoutCallback, nullptr);
+			k_timer_user_data_set(&source.mTimer, &source);
+		}
+
+		LOG_DBG("Registered a watchdog channel %d with feed interval %u ms", source.mChannel,
+			source.mFeedingTime);
+	}
+
+	if (startAutomatically) {
+		return Enable();
+	}
+
+	return true;
+}
+
+void WatchdogSource::Feed()
+{
+	if (sEnabled) {
+		if (mChannel == kInvalidChannel) {
+			LOG_ERR("Call for Watchdog FEED, but the watchdog has not been initialized");
+		}
+
+		wdt_feed(wdt, mChannel);
+	}
+}
+
+bool WatchdogSource::Install()
+{
+	if (sEnabled) {
+		return false;
+	}
+
+	/* Install a new timeout */
+	wdt_timeout_cfg wdtConfig = {};
+	/* Reset SoC when watchdog timer expires. */
+	wdtConfig.flags = WDT_FLAG_RESET_SOC;
+	/* The minimum window time must be 0 */
+	wdtConfig.window.min = 0;
+	/* The maximum window time must be equal for all channels */
+	wdtConfig.window.max = CONFIG_NCS_SAMPLE_MATTER_WATCHDOG_TIMEOUT;
+	/* The callback has no sense, because watchdog interrupt takes only ~64 us and CPU
+	 * cannot do anything useful */
+	wdtConfig.callback = nullptr;
+
+	mChannel = wdt_install_timeout(wdt, &wdtConfig);
+	if (-ENOMEM == mChannel) {
+		LOG_ERR("Cannot create a new Watchdog source because there is no more available Watchdog channels");
+		return false;
+	}
+
+	if (mChannel < 0) {
+		return false;
+	}
+
+	return true;
+}
+
+void WatchdogSource::TimerTimeoutCallback(k_timer *timer)
+{
+	if (sEnabled) {
+		WatchdogSource *watchdog = reinterpret_cast<WatchdogSource *>(k_timer_user_data_get(timer));
+		// Feeding using timer only of callback is set
+		if (watchdog && watchdog->mCallback) {
+			watchdog->mCallback(watchdog);
+		}
+	}
+}
+
+WatchdogSource::~WatchdogSource()
+{
+	/* Stop the global timer */
+	Disable();
+
+	sys_snode_t *node, *tmpNodeSafe;
+	SYS_SLIST_FOR_EACH_NODE_SAFE (&sWatchdogList, node, tmpNodeSafe) {
+		WatchdogSource *source = static_cast<WatchdogSource *>(node);
+		if (source == this) {
+			/* Remove this object from the list and stop its timer if it was enabled */
+			sys_slist_find_and_remove(&sWatchdogList, node);
```

Comment:
I guess both `sys_slist_find_and_remove()` and `k_timer_stop()` can be called unconditionally without checking if `this` is in the list.